### PR TITLE
Add .CopyTo() operator

### LINKED ISSUE #72: `.CopyTo()`
In order to improve allocations (ex: `ArrayPool.Rent()`) or to copy data to a specific pre-existing place, it may be useful to copy a sequence. Following APIs to be implemented:

```csharp
void CopyTo<T>(this IEnumerable<T> source, Span<T> span); 
void CopyTo<T>(this IEnumerable<T> source, IList<T> list);
void CopyTo<T>(this IEnumerable<T> source, IList<T> list, Index index);
```

Semantics:
* `void CopyTo<T>(this IEnumerable<T> source, IList<T> list);` will throw an exception if the `source` is longer than `array`. If `source` is shorter than `array`, then the remainder of `array` will be untouched.

### PR DESCRIPTION
**Pull request by viceroypenguin**
Fixes #72

### CODE REVIEW

**Comment by viceroypenguin** [comment]
@Arithmomaniac FYI: `TSource[]` version is provided because array has an equal implicit conversion to `IList<TSource>` and `Span<TSource>`, so there's no default function call for it.

**Comment by Arithmomaniac** [comment]
Will look at this tomorrow

Sent from phone. Please excuse brevity and typos.
________________________________

**Comment by Arithmomaniac** [comment]
@viceroypenguin anything to say on my PR suggestions?

**Comment by viceroypenguin** [comment]
> @viceroypenguin anything to say on my PR suggestions?

Which PR suggestions? I don't see any on this PR.

**Comment by Arithmomaniac** [comment]
Sorry, still gerrong used to GitHub I guess. Just posted

File: Source/SuperLinq/CopyTo.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+	/// This operator executes immediately.
+	/// </para>
+	/// </remarks>
+	public static void CopyTo<TSource>(this IEnumerable<TSource> source, IList<TSource> list, int index)
+	{
+		Guard.IsNotNull(source);
+		Guard.IsNotNull(list);
+		Guard.IsGreaterThanOrEqualTo(index, 0);
+
+		if (list is List<TSource> l)
```
[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Source/SuperLinq/CopyTo.cs:139]: JIT is optimized for `List<>` specifically vs `IList<>`, for up to 50% perf improvement when accessing `List<>`. For perf reasons, will be keeping separate.

File: Source/SuperLinq/CopyTo.cs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		}
+
+		var i = 0;
+		foreach (var el in source)
+		{
+			if (i > array.Length)
+				ThrowHelper.ThrowArgumentException(nameof(array), "Destination is not long enough.");
+
+			array[i++] = el;
+		}
```
[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Source/SuperLinq/CopyTo.cs:87]: Some `.CopyTo()` implementations only receive array, and not `Span<>`. Did update to take advantage of these in #81

File: Source/SuperLinq/CopyTo.cs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	/// <para>
+	/// All data from <paramref name="source"/> will be copied to <paramref name="span"/> if possible. If <paramref
+	/// name="source"/> is shorter than <paramref name="span"/>, then any remaining elements will be untouched. If
+	/// <paramref name="source"/> is longer than <paramref name="span"/>, then an exception will be thrown.
+	/// </para>
+	/// <para>
+	/// This operator executes immediately.
+	/// </para>
+	/// </remarks>
+	public static void CopyTo<TSource>(this IEnumerable<TSource> source, Span<TSource> span)
```
[the remark this answers is not in the document]
Comment (viceroypenguin) [reply Source/SuperLinq/CopyTo.cs:28]: Addressed in #81.

**Pull request merged**